File: 37-SudokuSolver/37-SudokuSolver.py
# Last updated: 31/8/2025, 5:29:33 pm
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        n = 9
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9

        emptyCells = []
        for i in range(n):
            for j in range(n):
                if board[i][j] == '.':
                    emptyCells.append((i, j))
        
        for i in range(n):
            for j in range(n):
                if board[i][j] == '.': continue
                cell = 1 << (int(board[i][j]) - 1)
                box = (i // 3) * 3 + (j // 3)
                rows[i] |= cell
                cols[j] |= cell
                boxes[box] |= cell
            
        # print([format(rows[i], '09b') for i in range(n)])
        # print([format(cols[i], '09b') for i in range(n)])
        # print([format(boxes[i], '09b') for i in range(n)])

        # print(emptyCells)
        ret = []
        def rec(index):
            if index == len(emptyCells):
                ret = [board[k][:] for k in range(n)]
                return True
            i, j = emptyCells[index]
            box = (i // 3) * 3 + (j // 3)
            for num in range(9):
                cell = 1 << num
                if rows[i] & cell == 0 and cols[j] & cell == 0 and boxes[box] & cell == 0:
                    board[i][j] = str(num + 1)
                    prev_row = rows[i]
                    prev_col = cols[j]
                    prev_box = boxes[box]
                    rows[i] |= cell
                    cols[j] |= cell
                    boxes[box] |= cell
                    if rec(index + 1): return True
                    rows[i] = prev_row
                    cols[j] = prev_col
                    boxes[box] = prev_box
            
            return False
        
        rec(0)
        return ret

File: 37-SudokuSolver/37-SudokuSolver.cpp
// Last updated: 21/7/2025, 3:33:54 pm
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<int> boxes(9, 0);
        vector<int> rows(9, 0);
        vector<int> cols(9, 0);
        vector<pair<int, int>> need;
        int n = 9;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == '.') {
                    need.push_back({i, j});
                } else {
                    int pos = 1 << (board[i][j] - '1');
                    int box = (i / 3) * 3 + (j / 3);
                    rows[i] |= pos;
                    cols[j] |= pos;
                    boxes[box] |= pos;
                }
            }
        }

        function<bool(int)> dfs = [&](int index) {
            if (index == need.size()) return true;

            int i = need[index].first;
            int j = need[index].second;
            int box = (i / 3) * 3 + (j / 3);

            for (int val = 0; val < 9; ++val) {
                int pos = 1 << val;
                if ((rows[i] & pos) == 0 && (cols[j] & pos) == 0 && (boxes[box] & pos) == 0) {
                    int r = rows[i];
                    int c = cols[j];
                    int b = boxes[box];

                    rows[i] |= pos;
                    cols[j] |= pos;
                    boxes[box] |= pos;
                    board[i][j] = char('1' + val);

                    if (dfs(index + 1)) return true;

                    rows[i] = r;
                    cols[j] = c;
                    boxes[box] = b;
                    board[i][j] = '.';
                }
            }
            return false;
        };

        dfs(0);
    }
};